lib/class-wp-rest-block-types-controller.php

```diff
@@ -0,0 +1,326 @@
+<?php
+/**
+ * REST API: WP_REST_Block_Types_Controller class
+ *
+ * @since      5.4.0
+ * @subpackage REST_API
+ * @package    WordPress
+ */
+
+/**
+ * Core class used to access block types via the REST API.
+ *
+ * @see   WP_REST_Controller
+ */
+class WP_REST_Block_Types_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = '__experimental';
+		$this->rest_base = 'block-types';
+	}
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @see   register_rest_route()
+	 */
+	public function register_routes() {
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<namespace>[a-zA-Z0-9_-]+)/(?P<name>[a-zA-Z0-9_-]+)',
+			array(
+				'args'   => array(
+					'name'      => array(
+						'description' => __( 'Block name', 'gutenberg' ),
```

Comment:
In the PHP, it is named name not slug. I don't think we should change this.

Reply:
Well, we should in the code example I shared to align :)

Reply:
The naming of name has been in WP core since 5.0 https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/class-wp-block-type.php#L24. It would not be easier to rename in PHP. Can it be changed in JS?

Reply:
I don’t think we have this issue on the client. It’s just the to I referenced.

Reply:
Seems like name is used in js as well. https://github.com/WordPress/gutenberg/blob/6ddcdc3d63884e81576c79528835d40719c91f9f/packages/block-library/src/paragraph/block.json#L2

Reply:
I think there might be two things discussed here.

`"core/paragraph"` is a block name, yes.

But if we're talking about just "paragraph"  within that, where block name is composed of a namespace, a forward slash, and an X, the question is about what to call X ? I don't know that I've come across too many references to this part alone to have a good sense of how it's called.

The [grammar document](https://github.com/WordPress/gutenberg/blob/master/docs/grammar.md) refers to the whole thing as "namespaced block name", and the parts as "block name part" (applying to both parts before and after the forward slash). The implementation of the parser calls it `name` ([source](https://github.com/WordPress/wordpress-develop/blob/9c3a5d30ab93448b19519b4adb3669b4a6f63150/src/wp-includes/class-wp-block-parser.php#L413)). Confusingly, the [block registration RFC](https://github.com/WordPress/gutenberg/blob/master/docs/rfc/block-registration.md#block-api) refers to "block name" as being composed of the namespace and "block name" (🤷).

Maybe it's fine to call it as block name, and if ever a distinction needs to be made, we can call the full string as "namespaced block name".